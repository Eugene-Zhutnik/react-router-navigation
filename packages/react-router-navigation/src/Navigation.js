/* @flow */

import * as React from 'react'
import { Route } from 'react-router'
import { CardStack, type CardsRendererProps } from 'react-router-navigation-core'
import type { NavigationProps, NavBarProps, NavigationHeaderProps } from './TypeDefinitions'
import DefaultNavigationRenderer from './DefaultNavigationRenderer'
import NavBar from './NavBar'

type Props = NavigationProps & {
  children?: React$Node,
}

class Navigation extends React.Component<Props> {
  renderHeader = (
    headerProps: NavBarProps<CardsRendererProps & NavigationHeaderProps> &
      CardsRendererProps &
      NavigationHeaderProps,
  ) => {
    const { cards, scene: { route } } = headerProps
    const activeCard = cards.find(card => card.path === route.routeName)
    const navBarProps = { ...headerProps, ...activeCard }
    if (navBarProps.hideNavBar) return null
    if (navBarProps.renderNavBar) {
      return navBarProps.renderNavBar(headerProps)
    }
    return <NavBar {...headerProps} />
  }

  render() {
    return (
      <Route>
        {({ history }) => (
          <CardStack
            {...this.props}
            history={history}
            render={cardsRendererProps => (
              <DefaultNavigationRenderer
                {...this.props}
                {...cardsRendererProps}
                renderHeader={this.renderHeader}
              />
            )}
          />
        )}
      </Route>
    )
  }
}

export default Navigation
