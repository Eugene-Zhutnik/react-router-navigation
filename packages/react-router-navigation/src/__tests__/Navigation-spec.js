import React from 'react'
import { Router, Route } from 'react-router'
import createHistory from 'history/createMemoryHistory'
import renderer from 'react-test-renderer'
import { componentFactory } from './utils'
import './__mocks__'
import Navigation from './../Navigation'

it('<Navigation /> renders correctly', () => {
  const history = createHistory()
  const component = renderer.create(
    <Router history={history}>
      <Navigation hideNavBar>
        <Route exact path="/" render={componentFactory('Index')} />
        <Route path="/hello" render={componentFactory('Hello')} />
      </Navigation>
    </Router>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<Navigation /> renders correctly with initialIndex and initialEntries prop ', () => {
  const history = createHistory({
    initialIndex: 1,
    initialEntries: ['/', '/hello', '/goodbye'],
  })
  const component = renderer.create(
    <Router history={history}>
      <Navigation hideNavBar>
        <Route exact path="/" render={componentFactory('Index')} />
        <Route path="/hello" render={componentFactory('Hello')} />
        <Route path="/goodbye" render={componentFactory('Goodbye')} />
      </Navigation>
    </Router>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<Navigation /> re-renders correctly when "push" action is called', () => {
  const history = createHistory()
  const component = renderer.create(
    <Router history={history}>
      <Navigation hideNavBar>
        <Route exact path="/" render={componentFactory('Index')} />
        <Route path="/hello" render={componentFactory('Hello')} />
      </Navigation>
    </Router>,
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  history.push('/hello')
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<Navigation /> re-renders correctly when "replace" action is called', () => {
  const history = createHistory({
    initialIndex: 1,
    initialEntries: ['/', '/hello'],
  })
  const component = renderer.create(
    <Router history={history}>
      <Navigation hideNavBar>
        <Route exact path="/" render={componentFactory('Index')} />
        <Route path="/hello" render={componentFactory('Hello')} />
        <Route path="/goodbye" render={componentFactory('Goodbye')} />
      </Navigation>
    </Router>,
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  history.replace('/goodbye')
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
