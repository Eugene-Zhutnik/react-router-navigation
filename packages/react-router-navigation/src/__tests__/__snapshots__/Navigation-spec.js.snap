// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> re-renders correctly when "push" action is called 1`] = `
<View>
  <View>
    <View>
      <Card>
        <Text>
          Index
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> re-renders correctly when "push" action is called 2`] = `
<View>
  <View>
    <View>
      <Card>
        <Text>
          Index
        </Text>
      </Card>
      <Card>
        <Text>
          Hello
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> re-renders correctly when "replace" action is called 1`] = `
<View>
  <View>
    <View>
      <Card />
      <Card>
        <Text>
          Hello
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> re-renders correctly when "replace" action is called 2`] = `
<View>
  <View>
    <View>
      <Card />
      <Card>
        <Text>
          Goodbye
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> renders correctly 1`] = `
<View>
  <View>
    <View>
      <Card>
        <Text>
          Index
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> renders correctly with initialIndex and initialEntries prop  1`] = `
<View>
  <View>
    <View>
      <Card />
      <Card>
        <Text>
          Hello
        </Text>
      </Card>
    </View>
  </View>
</View>
`;
